### ISSUE TITLE
[q-extract-torch] PyTorch Quantization Parameter extract

### ISSUE THREAD

**Issue by ssafyWooJeong** [Samsung/ONE#11551]
## What
Let's develop `PyTorch Quantization Parameter extract`, a tool to extract quantization parameters in Quantized PyTorch model, to be used by `q-implant`.
#11254 

## Why
* ONE has its own quantization tool, but it has limitations in applying latest quantization techniques implemented in PyTorch.
* There is `q-implant` module which can import quantization parameters into circle, in development.
* We will implement `q-extract-torch`, which can extract quantization parameters for `q-implant`
* Expected benefits
   
  *  The latest quantization techniques implemented in PyTorch can be applied to the circle model
  *  Circle model will have improved accuracy and performance by implanting extracted quantization parameters

## Tasks
* Define `q-extract-torch`'s input arguments.
* Define data format (quantization parameters + weights of the model).
* Implement `q-extract-torch`.
* Mapping names between PyTorch model and Circle model.
* Test `q-extract-torch`

## Structure
```
- q-extract-torch
  - test
    - example modules like res/PyTorchExamples/examples
  - include 
    - circle     → pics 
  - Torch2Circle.py   →   Convert PyTorch to Circle
  - Torch_Circle_Mapper.py    →   Generate mapping data between PyTorch and Circle by using Tensor data
  - Torch_Extractor.py     →   Extract/Save quantization parameters from PyTorch
  - Torch_Extract_Helper.py     →   For convenient use
  - Test_Ruuner.py
```

## Notes
It uses `tflite2circle` and `pics` to make name mapping between PyTorch and Circle.

**Comment by seanshpark** [comment]
this issue looks identical to #11254 . why did you add this?

**Comment by ssafyWooJeong** [comment]
#11254 needs qparam.json and numpy files to implant quantization parameters to circle.
This issue handles python script to export quantization parameters(qparam.json and numpy files) from a PyTorch Model.

**Comment by seanshpark** [comment]
1/ It's a bit hard to understand what this tool from reading the comment
2/ I've talked offline with @jinevening and IIUC , this tool is to 
- extract quantization data from pytorch model
- save the data for `q-implant` tool

thus this tool works with `q-implant` but is not `q-implant`
and using word `export` doesn't seem to be appropriate.
- center of operation for this is `q-implant`
- this tool will extract data from pytorch model
- this tool will save/convert data to `q-implant` tool

**Comment by jinevening** [comment]
@ssafyWooJeong Could you find a new name for the module and update title (replace `q-implant` with your module name)?

**Comment by ssafyWooJeong** [comment]
> * extract quantization data from pytorch model
> * save the data for `q-implant` tool

Yes that's purpose of the code, and I'm really sorry about my poor explanation.

And I understand appropriate naming is most important job, but hard to find/decide appropriate naming for it.
Do you thinks `q-extract` or `q-torch-extract` represent module's purpose? Or there are any better option?

@jinevening Ok, I will change the title when new name is fixed.

**Comment by seanshpark** [comment]
> Do you thinks q-extract or q-torch-extract represent module's purpose? Or there are any better option?

I think `q-extract-torch` would be OK. when some model is to be supported, say `lighter`,
we can make it like `q-extract-lighter`, so that `q-extract-` tools are listed near in directory list.

**Comment by ssafyWooJeong** [comment]
Thank you for your recommendation!

And we will make draft pr of `q-extract-torch` when skeleton of test is ready.

**Comment by seanshpark** [comment]
> Mapping names between PyTorch model and Circle model.

Can you describe more about this? Are there any limitations with operator types and maybe with other attributes?

**Comment by ssafyWooJeong** [comment]
Before explanation about Mapping, there are limitations.
Not tested and checked all operators yet, but I heard yesterday that there is not mapped operator(ConvTranspose) exists.
We plan to check operators on res/PyTorchExamples/examples.
If we found not mapped operator on draft code, we will closely look whether we can solve/map it or not.
<br/> 
@jinevening said that name disagreement between of PyTorch and Circle could be manually mapped (by human, as I understand). But mapping names as much as possible could reduce human effort, so we tried to map it.
Mapping has limitation and not mapped attribute could exists, we extract mapped quantization parameters to `qparam.json` and not mapped quantization parameters to 'not_mapped_qparam.json` to be manually mapped by human.


![Mapping Flow](https://github.com/Samsung/ONE/assets/136890801/0b0915eb-ed26-4a43-ac86-893598da9e80)  
It's the flow of mapping between PyTorch and Circle by using tensor value.  <br/> 
We process it under some assumptions.
* Converting same PyTorch model to Circle model (PyTorch → Onnx → TF → tflite → Circle) always same
  * When PyTorch model's name 'abc' converted 'def' on Circle, we can get `def` on another conversion of same model
* Converting PyTorch model to Circle model does not change tensor value
  * When finding same tensor value(corresponding tensor between PyTorch and Circle), we can get corresponding name.
  * Not supported operator can exists when tensor changed during conversion
* It rely on `Tensor`. If same Tensor(shape, value) exists, conflict occurs and should make exception
<br/>
- First, when you look at center of dashed line area (near Pink area in figure), make mapping between tensor value(more accurately hash of pytorch tensor's numpy) and pytorch name. (represented as `reverse_mapping` on figure)
- Second, when same hash value of circle tensor(like above, hash of circle tensor's numpy), make mapping between pytorch name and circle name (represented as `mapping` on figure)
- At a same time, make mapping between pytorch module name and circle tensor's index. (represented as `op_mapping`)
  - When pytorch's tensor name `abc.def.ghi` → `abc.def` would be key
  - If mapping value(set) belong to circle's operator's input set, remained attributes(which don't have value) on circle operator can be mapped.
- At bottom of dashed line area if pytorch model have activation, get previous module's name of it
  - Some activation(such as relu) don't have value itself. Just relying on mapping solely tensor can't map those activation.

**Comment by ssafyWooJeong** [comment]
Found problem about assumption while checking mapping...
Sorry about our faults...

We will check whether we can solve it or not...

**Comment by seanshpark** [comment]
About processing with `pytorch`.
We don't have direct tool to import pytorch model but have onnx model.
Is it possible to process with onnx model?

**Comment by 0minyoung0** [comment]
Thank you always for your comment.

The purpose of `q-implant` and `q-extract` is
to enable the use of the latest quantization algorithms that are not supported by the ONE framework.

We understand that the quantization supported by ONNX itself supports only traditional algorithms, 
and new algorithms have generally been released as PyTorch or TensorFlow.

When we checked directly whether we can convert quantized PyTorch model to ONNX model, 
it seems that there are models that convert to onnx, and there are models that do not convert.

If what we think is correct, the quantized model in ONNX is likely to have a traditional quantizer parameter, 
which is different from the purpose of `q-implant`.

Despite the above situation, 
I'm not sure it would be meaningful to implement (`q-extract-onnx?`) in a way that works on ONNX 
instead of implementing q-extract-torch.

We have a face-to-face meeting with @jinevening tomorrow at 1 p.m. 
We will decide whether we can proceed with the above in our project at the meeting.

**Comment by seanshpark** [comment]
Can someone draw a diagram with models and tools? for better understanding.
- models in circle
- tools in rectangle
- models: pth, onnx, circle, opt.circle, q.opt.circle, ...
   - pth: pytorch model
   - onnx: onnx model
   - circle: circle model (result of tflite2circle)
   - opt.circle: optimized circle model (result of circle2circle)
   - q.opt.circle: quantized optimized circle model (result of circle-quantizer)
- tools: one-import-onnx, circle2circle, circle-quantizer, q-implant, q-extract

**Comment by 0minyoung0** [comment]
Let me tell you the result of our meeting.

New quantize techniques are sometimes implemented using PyTorch.
Therefore, `q-implant` aims to output quantized circle model
by inputting `circle model` and [`json & npy files`](https://github.com/Samsung/ONE/issues/11254#issuecomment-1692780934) containing quantization parameters of the quantized PyTorch model.

However, converting quantized pytorch model to onnx model is not officially supported.
Therefore, starting q-implant from onnx rather than pytorch is difficult to satisfy `q-implant`'s design objectives.

`q-extract` is a module designed to obtain `circle model` and `json & npy files` corresponding to the input of q-implant.
Therefore, we judged that implementing `q-extract` for `quantized onnx model` rather than `quantized pytorch model`
was also not in line with the existing purpose.

![image](https://github.com/Samsung/ONE/assets/122426037/4f6a49e2-19ed-4082-b50d-2bad96899121)

The figure above is a diagram of q-extract-torch.

I will explain the inputs and outputs of the q-extract-torch module in text.
Input 1: `original PyTorch model` (not quantized)
Input 2: `quantized PyTorch model` (quantized)
Output 1: `Circle model` (not quantized, to be used as input for q-implant)
Output 2: `json & npy files` (to be used as input for q-implant)

In addition, I will explain the functionality of the `q-extract-torch` module.
First, Convert `original PyTorch model (Input 1)` to create `original Circle model (Output 1)`.
Second, `Two input PyTorch models (Input 1 & 2)` store `mapping information` for models before and after quantization.
Based on this `mapping information`, output `json & npy files (Output 2)` corresponding to the input of q-implant.

**Comment by seanshpark** [comment]
Thanks for the explanation. It's more understandable :)
Q) how many days are left to work on this item with SSAFY?
Q) how many pytorch operators do you plan to support in this period?
Q) @jinevening , do you have future plans for `q-extract-torch` ?

**Comment by jinevening** [comment]
Two weeks are left for SSAFY (maybe one-week working days considering Chu-seok). 10/6 is the final presentation day.

> Q) @jinevening , do you have future plans for q-extract-torch ?

I don't have any future plan for `q-extract-torch` for now. I'd like to see how this idea grows in draft.

**Comment by 0minyoung0** [comment]
> Q) how many pytorch operators do you plan to support in this period?

We apologize for not being able to provide a clear plan for the project right now. Ultimately, we want to support MobileNet, ResNet, and Inception sequentially. We are first trying to implement mapping on operators of MobileNet. 
However, supporting Batch Normalization is currently challenging. We are trying to figure out if we can resolve this issue. 
Once this problem is resolved, we will proceed with the support of the following models. However, if we don't solve the problem, we will record what operators we succeeded in mapping and failed during this project in terms of experience.

**Comment by ssafyWooJeong** [comment]
#11561 
Experimental report about `q-extract-torch` and `q-implant`.
At first, `q-extract-torch` is intended to support `q-implant` to produce `q-param.json` and `*.npy` files which are used in `q-implant`.
However it has major limitations and maybe relied on inaccurate assumption, so we just left our approaches.

We developed/tested `q-extract-torch`  on torchvision mobilenet_v2 and resnet18/50, so maybe it only works on CNN based models.
Also we only handle rank 4 as draft stage, it would not work properly (Permutation would not work.)

Below is our approaches and what we learned/thought.

- Names are changed due to different naming conventions on model conversion step. (PyTorch→Onnx→Tensorflow→Tensorflow Lite→Circle).
So name mapping between PyTorch↔Circle required.
- Most of Tensor values are kept during model conversion step.
So if we find corresponding Tensor value, we can map between PyTorch and Circle.
- At this point, we tried to map `PyTorch` and `Circle` based on Tensor value.
- However during model conversion step(such as PyTorch→Onnx), some values are changed due to optimization.
Mapping based on Tensor value have to solve difference between Original Tensor Value and Optimized Tensor value, so we have assigned/marked unique integer value.
(Even the value is set to integer, data type should be kept in float. Because there are operators supported float only)
When we track the marked value(even the little error occured on tensor, we can track it by rounding it. Because we have marked all tensors values to integer values), we can map PyTorch Tensor and Cirlce Tensor.
- When exporting `PyTorch` to `Onnx`, not existed value is added to onnx model. For example, even the Conv2d operator on PyTorch doesn’t have bias, correspoing Conv2d Onnx Operator have bias filled with zeros. So multiple tensors filled with zeros could exist.
- When the operator changed during model conversion, some of them should be handled by `q-extract-torch`.
For example, Batch Normalization is supported on PyTorch, Onnx and Tensorflow, but not supported by Tensorflow Lite. Tensorflow resolves Batch Normalization by split it into Mul and Add .
    - On Mapping Step, set PyTorch [`Batch Norm`](https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm2d.html) like below, can track/map PyTorch `BatchNorm` ↔ Circle `Mul` and `Add`
      - running_mean : 0
      - running_var : 1
      - eps : 0
      - momentum : 0  (used as 1 - momentum, so should be 0)
      - *γ(weight on operator) : Correspond to Mul Operator on tflite/circle*
      - *β(bias on operator) : Correspond to Add Operator on tflite/circle*
  - On Extraction, maually calculate Circle Mul/Add value from PyTorch BatchNorm. [(What Tensorflow→Tflite Conversion do)](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/toco/graph_transformations/resolve_batch_normalization.cc#L37)
    - mul : weight
    - add : bias -  running_mean * weight
- Durning Model conversion, some operators could be added (like [transpose](https://github.com/onnx/onnx-tensorflow/blob/ee0c5e537b3cebbddc5773871e6786e6468c7f3f/onnx_tf/handlers/backend/conv_mixin.py#L129)), `q-implant` should handle for operators which can be added on model conversion, and forward propagate the `q-param` for operators that would not change quantization parameters.(min/max range, zero point, scale)
- As I know, if the operator is not subclass of `torch.nn.Modules`, we can’t access the operator from the module. So we can’t get `q-param` for those operators which are not subclass of `torch.nn.Modules`. (Operator on `torch` like `add`, `mul`, `mean` and etc.) 
`q-implant` should handle those operators that can’t be extracted from PyTorch.

- We have some changes on `q-implant` to use extracted data from torchvision mobilenet/resnet. (What represented as `q-implant` should handle)
    - Assumed min/max with dtype, scale and zerop of operator (Because `q-param.json` format don't have min/max parameter)
    - Assumed operator's min/max by using inputs' quantization parameter, and calculated zero point, scale by using dtype's range.
    - For example, `Add` don't have quantization parameter because it is not subclass of `torch.nn.Modules`. Let inputs as x, y and output as z. We assumed `Add`'s quantization parameter as below.
 ```
z_min = min(x_min, y_min, x_min + y_min)
z_max = max(x_max, y_max, x_max + y_max)
```
  
- We think handling it well would work. (But It will possess errors.) However even it works in some parts, it would not be good solution because we have to handle those cases on one by one.(handle opcode differently)
- Handling those on `q-implant` would not be good option, and we have to find way to solve it on extract part.
When we considered it, there are two ways. (What we thought. And there could be other/better solutions)
    - 1) access and get quantization parameter of operators(of `torch` not `torch.nn.Modules`), map it corresponding circle name.
    - 2) Export both not quantized model(for mapping) and quantized model(for extraction) to `onnx`, and process them.
- However, as we know. (What we know could be wrong.)
    - 1) Not sure about there is way to access them from `torch.nn.Modules`
    - 2) Some quantized operator in `PyTorch` don't support `onnx` export.


- We don't know better way for `q-extract-torch` for now, those are what we have tried.
- Additionally, `PyTorch`'s `QuantStub` and `DeQuantStub` is not exported to `onnx`, so circle don't have quantize/dequantize operator when we run `q-implant`.
Maybe we have to add below code to add `q-implant`
```
  for (auto node : loco::input_nodes(g))
  {
    auto circle_node = loco::must_cast<luci::CircleNode *>(node);
    auto quantize = node->graph()->nodes()->create<luci::CircleQuantize>();
    quantize->name(circle_node->name() + "_Quantize");
    quantize->dtype(circle_node->dtype());
    quantize->rank(circle_node->rank());
    for (uint32_t i = 0; i < circle_node->rank(); ++i)
      quantize->dim(i).set(circle_node->dim(i).value());

    quantize->shape_status(luci::ShapeStatus::VALID);

    copy_quantparam(circle_node, quantize);
    circle_node->quantparam(nullptr);
    circle_node->dtype(loco::DataType::FLOAT32);

    loco::replace(circle_node).with(quantize);
    quantize->input(circle_node);
    luci::add_origin(quantize, luci::get_origin(circle_node));
  }
```
